### PR TITLE
Fix positions of qualified names

### LINKED ISSUE #763: Positions of qualified names
Currently, looks like namePos in AST is the start of the local name, even when there's a prefix.

This is not what is needed. It's the combination of the prefix and the local name that has meaning.  When an error refers to a qualified name, the entire qualified name should be underlined.

This means that:

- namePos should be the start of the prefix when there is one, not the local name
- we need to make sure error ranges referring to names cover the entire qualified name

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Fix positions of qualified names
Resolves #763

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front.syntax/token.bal [hunk 1]
```diff
@@ -551,13 +551,43 @@ public readonly class SourceFile {
...
+
     function scannedLines() returns readonly & ScannedLine[] => self.lines;
 
     function scannedLine(int lineNumber) returns ScannedLine {
         return self.lines[lineNumber - 1];
     }
 }
 
+function qualifiedIdentifierEndCodePointIndex(ScannedLine line, int startCodePointIndex) returns int {
+    var [fragIndex, fragmentIndex] = scanLineFragIndex(line, startCodePointIndex);
```
Comment (jclark) [inline compiler/modules/front.syntax/token.bal]: Space after var

File: compiler/modules/front.syntax/token.bal [hunk 2]
```diff
@@ -551,13 +551,43 @@ public readonly class SourceFile {
...
+    string[] fragments = line.fragments;
+    FragCode[] fragCodes = line.fragCodes;
+    int endCodePoint = startCodePointIndex;
```
Comment (jclark) [inline compiler/modules/front.syntax/token.bal]: Variable name should end with Index

File: compiler/modules/front.syntax/parseStmt.bal [hunk 3]
```diff
@@ -163,22 +163,21 @@ function finishIdentifierStmt(Tokenizer tok, string name, Position startPos, Pos
...
     Position endPos = tok.previousEndPos();
     if cur == "(" {
-        FunctionCallExpr expr = check finishFunctionCallExpr(tok, prefix, name, startPos, namePos);
+        FunctionCallExpr expr = check finishFunctionCallExpr(tok, prefix, name, startPos);
         return finishCallStmt(tok, expr, startPos);
     }
     else if cur == "." {
         Position opPos = tok.currentStartPos();
-        VarRefExpr varRef = { startPos, endPos:tok.previousEndPos(), name, namePos, prefix };
+        VarRefExpr varRef = { startPos, endPos:tok.previousEndPos(), name, qNamePos: startPos, prefix };
```
Comment (jclark) [inline compiler/modules/front.syntax/parseStmt.bal]: Space after endPos

File: compiler/modules/comm.diagnostic/diagnostic.bal [hunk 4]
```diff
@@ -27,6 +27,7 @@ public type File readonly & object {
     public function directory() returns string?;
     public function lineColumn(Position pos) returns LineColumn;
     public function lineContent(Position|Range range) returns [string, string, string];
+    public function qNameEndPos(Position startPos) returns Position;
```
Comment (jclark) [inline compiler/modules/comm.diagnostic/diagnostic.bal]: Probably better for this to be qNameRange and return a Range.

File: compiler/modules/front.syntax/parse.bal [hunk 5]
```diff
@@ -207,9 +207,19 @@ function parseError(Tokenizer tok, string? detail = ()) returns err:Syntax {
...
 }
 
-public function locationInDefn(ModuleLevelDefn defn, Position pos) returns d:Location {
-    return d:location(defn.part.file, pos);
+public function qNameLocationInDefn(ModuleLevelDefn defn, Position qnamePos) returns d:Location {
+    Position endPos = defn.part.file.qNameEndPos(qnamePos);
+    return locationInDefn(defn, qnamePos, endPos);
+}
+
+public function locationInDefn(ModuleLevelDefn defn, Position startPos, Position? endPos=()) returns d:Location {
```
Comment (jclark) [inline compiler/modules/front.syntax/parse.bal]: I would have thought this should take a Range.

**Review by jclark — CHANGES_REQUESTED** [review]
Generally looks good

File: compiler/modules/front.syntax/token.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/token.bal]: fixed

File: compiler/modules/front.syntax/token.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/token.bal]: fixed

File: compiler/modules/front.syntax/parseStmt.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/parseStmt.bal]: fixed

File: compiler/modules/comm.diagnostic/diagnostic.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/comm.diagnostic/diagnostic.bal]: fixed

File: compiler/modules/front.syntax/parse.bal (hunk 5 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/parse.bal]: fixed

File: compiler/modules/front.syntax/ast.bal [hunk 6]
```diff
@@ -262,7 +263,7 @@ public type MethodCallExpr record {|
     Position endPos;
     Position opPos; // position of .
     Position openParenPos;
-    Position namePos;
+    Position qNamePos;
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal]: No prefix here, so it's not a qname.

File: compiler/modules/front.syntax/token.bal [hunk 7]
```diff
@@ -446,7 +446,7 @@ class Tokenizer {
 
     function err(d:Message msg) returns err:Syntax {
         // XXX pass in endPos if we need to in order to be able to recreate the right endPos
-        return err:syntax(msg, loc=d:location(self.file, self.currentStartPos(), self.currentEndPos()));
+        return err:syntax(msg, loc=d:location(self.file, { startPos: self.currentStartPos(), endPos: self.currentEndPos() }));
```
Comment (jclark) [inline compiler/modules/front.syntax/token.bal]: Add currentRange method and use here.

File: compiler/modules/bir/bir.bal [hunk 8]
```diff
@@ -4,6 +4,7 @@ import wso2/nballerina.comm.diagnostic as d;
 
 public type SemType t:SemType;
 public type Position d:Position;
+public type Range d:Range;
```
Comment (jclark) [inline compiler/modules/bir/bir.bal]: Put this in verify.bal.

File: compiler/modules/front/codeGen.bal [hunk 9]
```diff
@@ -1635,7 +1643,7 @@ function validArgumentCount(CodeGenContext cx, bir:FunctionRef func, int nSuppli
 
 function genLocalFunctionRef(CodeGenContext cx, Environment env, string identifier, Position pos) returns bir:FunctionRef|CodeGenError {
     if !(lookup(identifier, env) == ()) {
-        return cx.unimplementedErr("local variables cannot yet have function type", pos);
+        return cx.unimplementedErr("local variables cannot yet have function type", cx.qNameRange(pos));
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Not a qname here

File: compiler/modules/front/codeGen.bal [hunk 10]
```diff
@@ -1653,8 +1661,8 @@ function genLocalFunctionRef(CodeGenContext cx, Environment env, string identifi
         else {
             msg = `${identifier} is not a function`;
         }
-        return cx.semanticErr(msg, pos);
-    }  
+        return cx.semanticErr(msg, cx.qNameRange(pos));
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Not a qname here

File: compiler/modules/front/codeGen.bal [hunk 11]
```diff
@@ -1688,7 +1696,7 @@ function getLangLibFunctionRef(CodeGenContext cx, bir:Operand target, string met
         string moduleName = t[0];
         bir:FunctionSignature? erasedSignature = getLangLibFunction(moduleName, methodName);
         if erasedSignature == () {
-            return cx.unimplementedErr(`unrecognized lang library function ${moduleName + ":" + methodName}`, pos);
+            return cx.unimplementedErr(`unrecognized lang library function ${moduleName + ":" + methodName}`, cx.qNameRange(pos));
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Comes from method call currently, so not a qname. But will in future, so pass in a Range.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/ast.bal (hunk 6 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/ast.bal]: fixed

File: compiler/modules/front.syntax/token.bal (hunk 7 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/token.bal]: fixed

File: compiler/modules/bir/bir.bal (hunk 8 above)
Comment (heshanpadmasiri) [reply compiler/modules/bir/bir.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 9 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 10 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/codeGen.bal (hunk 11 above)
Comment (heshanpadmasiri) [reply compiler/modules/front/codeGen.bal]: fixed

File: compiler/modules/front/const.bal [hunk 12]
```diff
@@ -19,9 +19,10 @@ type SimpleConst t:SingleValue;
 type FoldError ResolveTypeError;
 
 type FoldContext object {
-    function semanticErr(d:Message msg, s:Position pos, error? cause = ()) returns err:Semantic;
+    function semanticErr(d:Message msg, d:Position|d:Range pos, error? cause = ()) returns err:Semantic;
     // Return value of FLOAT_ZERO means shape is FLOAT_ZERO but value (+0 or -0) is unknown
     function lookupConst(string? prefix, string varName, d:Position pos) returns s:FLOAT_ZERO|t:OptSingleValue|FoldError;
+    function qNameRange(d:Position startPos) returns d:Range;
```
Comment (jclark) [inline compiler/modules/front/const.bal]: Are you using this? If not, then can delete here and in a couple of other places.

Comment (heshanpadmasiri) [reply compiler/modules/front/const.bal]: used only in `ConstFoldContext` (among subtypes) removed from others

**Pull request merged**